### PR TITLE
[subclasses] Do not fakeTensor const prop subclass args

### PR DESCRIPTION
**Pull request by IvanKobzarev**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #134855

The issue:

Const propagation checks only if arguments do not have FakeTensor. If argument is Subclass, it will pass this condition.

As a result Const Propogation execution happens without FakeTensorMode and having tensor factories inside Subclass.__torch_dispatch__ results that this Tensor is not Fakified.

Solution:

If we have subclasses arguments, do not count that const propagation is doable

### CODE REVIEW

File: test/functorch/test_aotdispatch.py [hunk 1]
```diff
@@ -6010,6 +6010,18 @@ def forward(self, x):
...
+        from torch.testing._internal.test_subclasses import SubclassWithTensorFactory
+
+        inp = SubclassWithTensorFactory(torch.zeros(3, 5))
+
+        def fn(x):
+            return 2 * x
+
+        ref_out = fn(inp)
+        out = torch.compile(fn, backend="aot_eager", fullgraph=True)(inp)
+        # self.assertEqual(ref_out, out)
```
Comment (zou3519) [inline test/functorch/test_aotdispatch.py:6023]: what's up here?

File: torch/testing/_internal/test_subclasses.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
+# mypy: ignore-errors
+
+import torch
+import torch.utils._pytree as pytree
+
+
+class SubclassWithTensorFactory(torch.Tensor):
```
Comment (zou3519) [inline torch/testing/_internal/test_subclasses.py:7]: this file should go in https://github.com/pytorch/pytorch/blob/main/torch/testing/_internal/common_subclass.py

**Review by zou3519 — DISMISSED** [review]
avoiding constant prop on subclasses seems reasonable to me. Left some minor comments, please read

**Review by zou3519 — COMMENTED** [review]
Sorry, I changed my mind. It's weird to me that Tensor subclasses are showing up in FakeTensor's `__torch_dispatch__`, I would expect that it never sees subclasses

File: test/functorch/test_aotdispatch.py (hunk 1 above)
Comment (IvanKobzarev) [reply test/functorch/test_aotdispatch.py:6023]: Yeah, I commented it, forgot to remove. 
Assert fails on some op that is not implemented in my Testing Subclass.

**Review by zou3519 — APPROVED** [review]
JK, this is FakeTensorMode

File: test/functorch/test_aotdispatch.py (hunk 1 above)
Comment (IvanKobzarev) [reply test/functorch/test_aotdispatch.py]: Resolved, changed impl fn to * ones only for float32

File: torch/_subclasses/fake_tensor.py [hunk 3]
```diff
@@ -1736,6 +1737,7 @@ def _dispatch_impl(
             should_allow_numbers_as_tensors(func)
             and not has_symbolic_sizes
             and not flat_arg_fake_tensors
+            and not has_subclasses
```
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/_subclasses/fake_tensor.py:1740]: Yeah, I also agree this is a bandaid fix and at least a comment saying how to do it properly would be good.

**Pull request closed without merging**